src/Microsoft.ML.Data/Transforms/SelectColumnsTransform.cs

```diff
@@ -0,0 +1,632 @@
+            {
+                get
+                {
+                    _host.Assert(0 <= colIndex && colIndex < _outputToInputMap.Length);
+                    return _outputToInputMap[colIndex];
+                }
+            }
+
+            internal Mapper(SelectColumnsTransform transform, ISchema inputSchema)
+            {
+                _transform = transform;
+                _host = transform._host.Register(nameof(Mapper));
+                _inputSchema = Runtime.Data.Schema.Create(inputSchema);
+
+                var selectedColumns = _transform.SelectColumns;
+                var keepHidden = _transform.KeepHidden;
+                _outputToInputMap = BuildOutputToInputMap(selectedColumns, keepHidden, _inputSchema);
+                _outputSchema = GenerateOutputSchema(selectedColumns, keepHidden, _inputSchema);
+            }
+
+            private static int[] BuildOutputToInputMap(IEnumerable<string> selectedColumns,
+                                                                             bool keepHidden,
+                                                                             Schema inputSchema)
+            {
+                var outputToInputMapping = new List<int>();
+                var columnCount = inputSchema.ColumnCount;
+                int outputIdx = 0;
+
+                for (int colIdx = 0; colIdx < columnCount; ++colIdx)
+                {
+                    if (!keepHidden && inputSchema.IsHidden(colIdx))
+                        continue;
+
+                    var columnName = inputSchema[colIdx].Name;
+                    if (selectedColumns.Contains(columnName))
+                    {
+                        outputToInputMapping.Add(colIdx);
+                        outputIdx++;
+                    }
+                }
+
+                return outputToInputMapping.ToArray();
+            }
+
+            private static Schema GenerateOutputSchema(IEnumerable<string> selectedColumns,
+                                                        bool keepHidden,
+                                                        Schema inputSchema)
+            {
+                var schemaColumns = new List<Schema.Column>();
+                var columnCount = inputSchema.ColumnCount;
+                for (int colIdx = 0; colIdx < columnCount; ++colIdx)
+                {
+                    if (!keepHidden && inputSchema.IsHidden(colIdx))
+                        continue;
+
+                    var column = inputSchema[colIdx];
+                    if (selectedColumns.Contains(column.Name))
+                    {
+                        schemaColumns.Add(column);
+                    }
+                }
+                return new Schema(schemaColumns);
+            }
+        }
+
+        private sealed class Row : IRow
+        {
+            private readonly Mapper _mapper;
+            private readonly IRow _input;
+            public Row(IRow input, Mapper mapper, Schema outputSchema)
```

Comment:
>Schema outputSchema [](start = 50, length = 19)

Looks weird

Reply:
Removed the outputSchema

---
In reply to: [225345828](https://github.com/dotnet/machinelearning/pull/1269#discussion_r225345828) [](ancestors = 225345828)